File: leetcode/copy-list-with-random-pointer/linked-list.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
  RandomListNode *copyRandomList(RandomListNode *head) {
    if (NULL == head){
      return NULL;
    }

    for (RandomListNode *cur = head; NULL != cur; cur = cur->next->next){
      RandomListNode *tmp = new RandomListNode(cur->label);
      tmp->next = cur->next;
      cur->next = tmp;
    }

    for (RandomListNode *cur = head; NULL != cur; cur = cur->next->next){
      if (cur->random){
        cur->next->random = cur->random->next;
      }
    }

    RandomListNode *newHead = head->next;
    RandomListNode *oldTail = head, *newTail = newHead;
    for (RandomListNode *cur = newHead->next; NULL != cur; cur = cur->next->next){
      oldTail->next = cur;
      oldTail = oldTail->next;

      newTail->next = cur->next;
      newTail = newTail->next;
    }
    oldTail->next = NULL;

    return newHead;
  }
};

File: leetcode/remove-nth-node-from-end-of-list/linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if nil == head || n <= 0 { return nil }

	dummyNode := &ListNode{0, head}

	goAhead := dummyNode
	cnt := 0
	for ; cnt <= n && nil != goAhead; cnt++ { goAhead = goAhead.Next }
	if cnt <= n { return nil }

	prev := dummyNode
	for ; nil != goAhead; { prev, goAhead = prev.Next, goAhead.Next }

	prev.Next = prev.Next.Next
	return dummyNode.Next
}
